plutus-contract/test/Spec/TxConstraints/MustProduceAtLeast.hs

```diff
@@ -0,0 +1,239 @@
+lessThanScriptBalance :: TestTree
+lessThanScriptBalance =
+    let amt = Ada.lovelaceValueOf (scriptBalance -500)
+        contract = mustProduceAtLeastContract amt amt noTokens $ mockWalletPaymentPubKeyHash w1
+    in  checkPredicateOptions
+            defaultCheckOptions
+            "Successful use of mustProduceAtLeast below script's balance"
+            (assertValidatedTransactionCount 2)
+            (void $ trace contract )
+
+-- Uses onchain and offchain constraint mustProduceAtLeast to spend entire token balance locked by the script
+spendTokenBalanceFromScriptAddress :: TestTree
+spendTokenBalanceFromScriptAddress =
+    let token = someTokens 1
+        contract = mustProduceAtLeastContract (scriptBalanceValue <> token) (scriptBalanceValue <> token) token $ mockWalletPaymentPubKeyHash w1
+        options = defaultCheckOptions
+            & changeInitialWalletValue w1 (token <>)
+    in  checkPredicateOptions
+            options
+            "Successful use of mustProduceAtLeast sending tokens"
+            (assertValidatedTransactionCount 2)
+            (void $ trace contract)
+
+-- Uses onchain and offchain constraint mustProduceAtLeast to spend more than the ada balance locked by the script, excess is paid by own wallet.
+higherThanScriptBalance :: TestTree
+higherThanScriptBalance =
+    let amt = Ada.lovelaceValueOf (scriptBalance + 5_000_000)
+        contract = mustProduceAtLeastContract amt amt noTokens $ mockWalletPaymentPubKeyHash w1
+    in  checkPredicateOptions
+            defaultCheckOptions
+            "Validation pass when mustProduceAtLeast is greater than script's balance"
+            (assertValidatedTransactionCount 2)
+            (void $ trace contract)
+
+-- Uses onchain and offchain constraint mustProduceAtLeast to spend more than the ada balance locked by the script. Other wallet's pubkeyhash is provided so cannot pay the excess, asserts contract error.
+  -- FAILS because it successfully gets funds from pubkey, shouldn't ownPaymentPubKeyHash only work for w1, not other wallet?
+higherThanScriptBalanceWithOtherWalletAsOwnPubkeyLookup :: TestTree
+higherThanScriptBalanceWithOtherWalletAsOwnPubkeyLookup =
+    let amt = Ada.lovelaceValueOf (scriptBalance + 5_000_000)
+        contract = mustProduceAtLeastContract amt amt noTokens $ mockWalletPaymentPubKeyHash w6
+    in  checkPredicateOptions
+            defaultCheckOptions
+            "Fail validation when mustProduceAtLeast is greater than script's balance and wallet's pubkey is not in the lookup"
+            (assertContractError contract (Trace.walletInstanceTag w1) (\case { ConstraintResolutionContractError Constraints.OwnPubKeyMissing -> True; _ -> False }) "failed to throw error"
+            .&&. assertValidatedTransactionCount 1)
+            (void $ trace contract)
+
+-- Contract error is thrown when there are not enough tokens  script and own wallet
```

Comment:
Don't understand the comment "not enough tokens script and own wallet". I think some word is missing.